plugins/node/opentelemetry-instrumentation-ioredis/src/ioredis.ts

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type * as ioredisTypes from 'ioredis';
+import {
+  InstrumentationBase,
+  InstrumentationConfig,
+  InstrumentationNodeModuleDefinition,
+  isWrapped,
+} from '@opentelemetry/instrumentation';
+import { IORedisInstrumentationConfig } from './types';
+import { traceConnection, traceSendCommand } from './utils';
+import { VERSION } from './version';
+
+export class IORedisInstrumentation extends InstrumentationBase<
+  typeof ioredisTypes
+> {
+  static readonly DB_SYSTEM = 'redis';
+  readonly supportedVersions = ['>1 <5'];
+
+  constructor(protected _config: IORedisInstrumentationConfig = {}) {
+    super('@opentelemetry/instrumentation-ioredis', VERSION, _config);
+  }
+
+  public setConfig(
+    config: IORedisInstrumentationConfig & InstrumentationConfig = {}
+  ) {
+    this._config = Object.assign({}, config);
+  }
+
+  init(): InstrumentationNodeModuleDefinition<typeof ioredisTypes>[] {
+    return [
+      new InstrumentationNodeModuleDefinition<typeof ioredisTypes>(
+        'ioredis',
+        this.supportedVersions,
+        moduleExports => {
+          this._logger.debug('Applying patch for ioredis');
+          if (isWrapped(moduleExports.prototype.sendCommand)) {
+            this._unwrap(moduleExports.prototype, 'sendCommand');
+          }
+          this._wrap(
+            moduleExports.prototype,
+            'sendCommand',
+            this._patchSendCommand()
+          );
+          if (isWrapped(moduleExports.prototype.connect)) {
```

Comment:
I'm not sure if/why this is needed. I added it because I saw it in grpc plugin: https://github.com/open-telemetry/opentelemetry-js/blob/b8afb044ff3e4bef8d9e375317f2501e82d1ff1e/packages/opentelemetry-instrumentation-grpc/src/grpc.ts#L82-L84

Reply:
Maybe @vmarchaud can explain the motivation there, but in general I've seen other plugins do this too. I think it's not needed, but some plugin authors have added checks like this because it feels like it would be safer. It is worth mentioning that because we use shimmer, we wouldn't be able to double wrap anyways and this behavior would be the same I think. Do the tests pass without these checks?

Reply:
@dyladan Pretty sure you can double wrap with shimmer (at least thats what i observed back when i migrated the http instrumentation) and thats i always added those checks. I just checked, there is just [this condition](https://github.com/open-telemetry/opentelemetry-js/blob/master/packages/opentelemetry-instrumentation/src/platform/node/instrumentation.ts#L123) that should avoid this issue but thats not what i found, it may be a bug ?

Reply:
you can wrap and unwrap with shimmer successfully many times as long as between wrap and unwrap nothing else will call wrap. With new instrumentation you can enable plugin, disable it, change something in config and the enable it again. I hope we could rethink again of replacing the shimmer with something I have proposed some months ago which allows safe wrapping without worrying if there was anything in the middle. And because all instrumentations are using the internal wrap methods the change will be easy.

Reply:
In the tests `isWrapped` returns always return `false` so from this point of view it could be removed.

The issue which came in my mind once seeing this was a different one: Consider someone else (e.g. some APM tool) wrapped the same API also using shimmer - in this case we would remove this tool.

Anyway, for now it seems these `if (isWrapped(` checks are used everywhere and if it turns out that this is problematic we should correct it everywhere instead of creating inconsistent instrumentations.

Another requirement for a safe unwrap/wrap sequence is that noone holds a reference to some of the old wrapped functions.

Reply:
I think that we might refactor this later to unwrap automatically when user tries to wrap it - inside instrumentation package.